### ISSUE TITLE
False warning about uninitialized field [-Wuninitialized] when calling static methods

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#13499]
|  |  |
| --- | --- |
| Bugzilla Link | [13127](https://llvm.org/bz13127) |
| Version | trunk |
| OS | Linux |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor,@zygoloid |

## Extended Description 
I get a lot of warnings when building Qt code with a current (revision 158529 if I read my git log output correctly) clang compiler.

<<< snip >>>
../../../../qt4/include/QtCore/../../../../src/qt4/src/corelib/tools/qsharedpointer_impl.h:595:43: warning: field is uninitialized when used here [-Wuninitialized]
    inline QWeakPointer(X *ptr) : d(ptr ? d->getAndRef(ptr) : 0), value(ptr)
                                          ^
1 warning generated.

At first glance the warning is correct, but d is defined as this:

Data d;

The struct data has a static methed getAndRef(ptrType).

So d->getAndRef(ptr) is calling a static method which admittedly is a unusual way to write this, but AFAIK it is legal to do so.

The warnings go away when changing qsharedpointer_impl.h to read:
inline QWeakPointer(X *ptr) : d(ptr ? Data::getAndRef(ptr) : 0), value(ptr)

**Comment by wheatman** [comment]
The warning no longer appears in this case as of 3.6
https://godbolt.org/z/MWWh4YGMd

code
```
class Foo
{
public:
static Foo* foo() { return 0; }
};

class Bar
{
public:
Bar() : f(f->foo()) {}
private:
Foo *f;
};
```

old warning 
```
<source>:10:11: warning: field 'f' is uninitialized when used here [-Wuninitialized]
Bar() : f(f->foo()) {}
          ^
1 warning generated.
Compiler returned: 0
```

**Comment by llvmbot** [comment]
@llvm/issue-subscribers-clang-frontend
